### PR TITLE
[eclipse/xtext#1777] ported xtend code to java

### PR DESCRIPTION
**Pull request by cdietrich**
[eclipse/xtext#1777] ported xtend code to java
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/model/TypeReference.java [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+	public TypeReference(String packageName, String className, List<TypeReference> arguments) {
+		if (packageName == null) {
+			throw new IllegalArgumentException("Invalid package name: " + packageName);
+		}
+		if (className == null) {
+			throw new IllegalArgumentException("Invalid class name: " + className);
+		}
+		this.packageName = packageName;
+		this.simpleNames = Arrays.asList(className.split("\\."));
```
Comment (nbhusare) [inline org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/model/TypeReference.java:105]: `this.` not needed here, and on the lines 109, 126, 127, 131, and 134.

Comment (cdietrich) [reply org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/model/TypeReference.java:105]: i kept it for consistency with place right next to it where it is needed cause of name clash with parameter

File: org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+				return;
+			case "xcore":
+				IResourceServiceProvider resourceServiceProvider3 = IResourceServiceProvider.Registry.INSTANCE
+						.getResourceServiceProvider(loadedResource);
+				if (resourceServiceProvider3 == null) {
+					try {
+						Class<?> xcore = Class.forName("org.eclipse.emf.ecore.xcore.XcoreStandaloneSetup");
+						xcore.getDeclaredMethod("doSetup", new Class[] {}).invoke(null);
+					} catch (ClassNotFoundException e) {
+						LOG.error("Couldn't initialize Xcore support. Is it on the classpath?");
```
Comment (nbhusare) [inline org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java:98]: "Couldn't initialize Xcore support. Is it on the classpath?"
You might like to change the log message to the following - Couldn't initialize Xcore support. It is possible that the required library is missing on the classpath.

File: org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+				return;
+			case "ecore":
+				IResourceServiceProvider resourceServiceProvider2 = IResourceServiceProvider.Registry.INSTANCE
+						.getResourceServiceProvider(loadedResource);
+				if (resourceServiceProvider2 == null) {
+					try {
+						Class<?> ecore = Class.forName("org.eclipse.xtext.ecore.EcoreSupportStandaloneSetup");
+						ecore.getDeclaredMethod("setup", new Class[] {}).invoke(null);
+					} catch (ClassNotFoundException e) {
+						LOG.error("Couldn't initialize Ecore support. Is 'org.eclipse.xtext.ecore' on the classpath?");
```
Comment (nbhusare) [inline org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java:83]: "Couldn't initialize Ecore support. Is 'org.eclipse.xtext.ecore' on the classpath?" -> Couldn't initialize Ecore support. It is possible that the library "org.eclipse.xtext.ecore" is missing from the project classpath.

Comment (cdietrich) [reply org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java:83]: did not want to refactor messages here

File: org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java (hunk 2 above)
Comment (cdietrich) [reply org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java:98]: did not want to refactor messages here

File: org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+				GenModelPackage.eINSTANCE.getEFactoryInstance();
+				IResourceServiceProvider resourceServiceProvider = IResourceServiceProvider.Registry.INSTANCE
+						.getResourceServiceProvider(loadedResource);
+				if (resourceServiceProvider == null) {
+					try {
+						Class<?> genModelSupport = Class.forName("org.eclipse.emf.codegen.ecore.xtext.GenModelSupport");
+						Object instance = genModelSupport.getDeclaredConstructor().newInstance();
+						genModelSupport.getDeclaredMethod("createInjectorAndDoEMFRegistration").invoke(instance);
+					} catch (ClassNotFoundException e) {
+						LOG.debug(
```
Comment (nbhusare) [inline org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java:68]: 1) For CNF, an error message is logged on lines 83, and 98. Here, a debug message is logged
2) Just to be consistent with other log messages, you might like to change it to - Couldn't initialize Ecore support. It is possible that the library "org.eclipse.xtext.ecore" is missing from the project classpath.

File: org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextDirectoryCleaner.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+				.transform(IXtextGeneratorFileSystemAccess::getPath).filter(it -> new File(it).isDirectory()));
+		Iterables.addAll(directories, extraDirectories);
+		DirectoryCleaner delegate = new DirectoryCleaner();
+		delegate.setUseDefaultExcludes(useDefaultExcludes);
+		excludes.forEach(it -> delegate.addExclude(it));
+		for (String it : directories) {
+			try {
+				delegate.cleanFolder(it);
+			} catch (FileNotFoundException e) {
+				throw Exceptions.sneakyThrow(e);
```
Comment (nbhusare) [inline org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextDirectoryCleaner.java:74]: I assume `Exceptions.sneakyThrow(e);` was used intentionally here.

File: org.eclipse.xtext.tests/src/org/eclipse/xtext/resource/AbstractXtextResourceSetTest.java [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+import java.io.File;
+
+import org.eclipse.emf.common.util.URI;
+import org.junit.Test;
+
+public abstract class AbstractXtextResourceSetTest extends AbstractResourceSetTest {
+	
+	@Override
+	  protected abstract XtextResourceSet createEmptyResourceSet();
```
Comment (szarnekow) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/resource/AbstractXtextResourceSetTest.java:21]: Indentation mixed up?

File: org.eclipse.xtext.tests/src/org/eclipse/xtext/resource/containers/LiveShadowedChunkedContainerTest.java [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+	public void setUp() throws Exception {
+		workspaceConfig = new WorkspaceConfig();
+		fooProject = new ProjectConfig("foo", workspaceConfig);
+		barProject = new ProjectConfig("bar", workspaceConfig);
+		XtextResourceSet rs0 = resourceSetProvider.get();
+		fooURI = IterableExtensions.head(fooProject.getSourceFolders()).getPath().trimSegments(1)
+				.appendSegment("foo.livecontainertestlanguage");
+		barURI = IterableExtensions.head(barProject.getSourceFolders()).getPath().trimSegments(1)
+				.appendSegment("bar.livecontainertestlanguage");
+		Map<String, ResourceDescriptionsData> chunks = Collections.unmodifiableMap(CollectionLiterals.newHashMap(
```
Comment (szarnekow) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/resource/containers/LiveShadowedChunkedContainerTest.java:90]: I'm not a big fan of this `newHashMap(Pair(..))` stuff... given this is only a test, I'm ok with it though.

File: org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextDirectoryCleaner.java (hunk 5 above)
Comment (szarnekow) [reply org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextDirectoryCleaner.java:74]: Yes, your assumption is correct. We could consider throwing a RuntimeIOException though it wouldn't change much.

**Review by szarnekow — APPROVED** [review]

**Pull request merged**